### DISCUSSION TITLE
vector and tensor features in MACE descriptor

### DISCUSSION THREAD

**Discussion by merlin021** [ACEsuit/mace#706] [Q&A] [answered]
Hi, I know we can get **equivariant features** by setting "invariants_only=False" in "get_descriptors", but is it possible to further **distinguish vector and tensor features** (both are equivariant features as far as I know) within it? I greatly appreciate it if such functions could be added. Thank you all!

**Comment by RokasEl** [comment, marked as answer]
Hi,

Yes, you can distinguish the individual components of `get_descriptors`. Here's an example breakdown:

If you use a model with `--max_L=2`, `--num_channels=64`, and two interaction layers, setting `invariants_only=False` results in a 640-dimensional output vector per atom. The structure of this vector is as follows:

- **First 64 dimensions:** Invariant descriptors from the first interaction block.  
- **Next 3 × 64 dimensions:** \( L = 1 \)  features from the first interaction block.  
- **Next 5 × 64 dimensions:** \( L = 2 \)  features from the first interaction block.  
- **Final 64 dimensions:** Invariant descriptors from the second interaction block.  

**Note:** Equivariant features \( L > 0 \) are not created for the final interaction layer.

Regarding your request for extracting specific \( L = x \) descriptors: while implementing such functionality in `get_descriptors` is feasible, it might be too niche to integrate directly. A better approach might be to write a custom postprocessing script to handle the extracted features.

Let me know if you need further clarification or guidance!

**Reply by merlin021** [reply]
Hey RokasEl,

Thank you so much for those super helpful information! May I ask why do we have 5 tensors?
Just wondering do we assume the tensor matrix is in shape of (3,3)? If that's the case, thinking we would have 6 or 9 tensors?
6 -> number of individual components of upper triangular tensors 
9 -> total individual components of the full tensor matrix

greatly appreciate that if you could share more information on this, thank you!

**Reply by RokasEl** [reply]
Sorry, in the original answer I was a bit overeager and identified the L=2 features as "tensors". That’s not entirely accurate—they are the \( L=2 \) components of a full rank-2 tensor (I've updated the original answer to reflect this).

The term "tensor" can be a bit overloaded here. MACE uses **spherical tensors**, indexed by $L$ and $m$ (with $m$ running from $-L$ to $L$), rather than **cartesian tensors** indexed by $(i, j, k, \dots)$. Cartesian and spherical tensors are related through a basis transformation, but this mapping is not unique.

For example, a rank-2 cartesian tensor can be decomposed as:

$$
X = \frac{1}{3} \text{Tr}(X) \cdot I + \frac{1}{2} (X - X^T) + \frac{1}{2} \left( X + X^T - \frac{2}{3} \text{Tr}(X) \cdot I \right)
$$

- $L=0$: Scalar ( $\text{Tr}(X)$ ).  
- $L=1$: Skew-symmetric part (3 components).  
- $L=2$: Traceless symmetric part (5 components).

This explains why there are 5 components for $L=2$ in MACE.

For more details, check these references:  
[arXiv:2306.06482](http://arxiv.org/abs/2306.06482)  
[arXiv:1807.02547](https://arxiv.org/abs/1807.02547)

**Reply by merlin021** [reply]
Hey RokasEl,

sorry for the late response. I think I got it super clear now! 
Thank you so much for the comprehensive explication of the ideas and refence paper too.  I'd rather try to convert those spherical tensors to cartesian tensors now by referring to the decomposition equation. Greatly appreciate for your help again, cheers!!!

**Comment by gabor1** [comment]
Can this structure info please be ported into the documentation (if it's not there already!)

**Comment by ilyes319** [comment]
It is in some form: 
![image](https://github.com/user-attachments/assets/8486d841-e514-46f6-a2fb-b4c1cd0b7daa)
Maybe we can give a notebook example on how to use descriptors. That would be a good addition.
